packages/liveblocks-core/src/__tests__/_utils.ts

```diff
@@ -50,94 +56,114 @@ function deepClone<T extends Json>(items: T): T {
   return JSON.parse(JSON.stringify(items)) as T;
 }
 
-export class MockWebSocket implements WebSocket {
-  CONNECTING = 0;
-  OPEN = 1;
-  CLOSING = 2;
-  CLOSED = 3;
+type Listener = (ev: IWebSocketEvent) => void;
+type CloseListener = (ev: IWebSocketCloseEvent) => void;
+type MessageListener = (ev: IWebSocketMessageEvent) => void;
 
-  static instances: MockWebSocket[] = [];
+export class MockWebSocket {
+  readonly CONNECTING = 0;
+  readonly OPEN = 1;
+  readonly CLOSING = 2;
+  readonly CLOSED = 3;
 
-  isMock = true;
+  readonly url: string;
 
-  callbacks = {
-    open: [] as Array<(event?: WebSocketEventMap["open"]) => void>,
-    close: [] as Array<(event?: WebSocketEventMap["close"]) => void>,
-    error: [] as Array<(event?: WebSocketEventMap["error"]) => void>,
-    message: [] as Array<(event?: WebSocketEventMap["message"]) => void>,
-  };
+  #readyState: number;
+  readonly sentMessages: string[] = [];
 
-  sentMessages: string[] = [];
-  readyState: number;
+  readonly #closeListeners: CloseListener[] = [];
+  readonly #errorListeners: Listener[] = [];
+  readonly #messageListeners: MessageListener[] = [];
+  readonly #openListeners: Listener[] = [];
 
-  constructor(
-    public url: string,
-    private onSend: (message: string) => void = () => {}
-  ) {
-    this.readyState = this.CONNECTING;
-    MockWebSocket.instances.push(this);
+  constructor(url: string = "ws://ignored") {
+    this.url = url;
+    this.#readyState = this.CONNECTING;
   }
 
-  close(_code?: number, _reason?: string): void {
-    this.readyState = this.CLOSED;
+  //
+  // WEBSOCKET API
+  //
+
+  get readyState(): number {
+    return this.#readyState;
   }
 
-  addEventListener<T extends "open" | "close" | "error" | "message">(
-    event: T,
-    callback: (event: WebSocketEventMap[T]) => void
-  ) {
-    this.callbacks[event].push(callback as any);
+  addEventListener(event: "message", listener: MessageListener): void; // prettier-ignore
+  addEventListener(event: "close", listener: CloseListener): void; // prettier-ignore
+  addEventListener(event: "open" | "error", listener: Listener): void; // prettier-ignore
+  // prettier-ignore
+  addEventListener(event: "open" | "close" | "message" | "error", listener: Listener | MessageListener | CloseListener): void {
+    if (event === "open") {
+      this.#openListeners.push(listener as Listener);
```

Comment:
Is there a clean way to avoid the "as" ?

I can think of one way, but it's a lot of code for little benefit.

```function isCloseListener(event: string; listener: SomeListener | AnotherListener):  is CloseListener {
  return event === "close"
}```

and then your if statement is ```if (isCloseListener(event, listener)) {
...
}```

Reply:
Unfortunately, type guards [are just as unsafe](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDAzgZSgJxmA5gCgDcBDAGxAFMAuRcAazDgHcwBKGk8i5FRFLHLkQBvAFCJEmClBCYkUAJ4AHCnGCJOlRAF5diAORgQAWwBGFTPoDcEgPS3EgXg3AEHuiAvqNEQE-RAA8aekYWHUQAFgAmK1EYdXxUDGw8fD9WVhFxCX8AOn5iTCgUAHUYKAALfH1iUwh9dMR7RBLSUkRgYhhW4ihJcFhjCndRIA) as an `as` would be here. TypeScript happily assume you know what you're doing in both approaches, and it's on us to keep these claims accurate 😬

I dislike `as`-expressions and always try to avoid them, but at the very least, an `as`-expression will fail if you're stating something that would be clearly nonsense. This is one of the reasons why I dislike complicated function overloads.

Here, the `as`-expressions are only used to select which union member applies to each `event` param, like the three overloads dictate should already be the case.

If typeguards would be actually typesafe to use, I would agree with your comment!